### PR TITLE
convert vault auth code from busl to mpl libraries

### LINKED ISSUE #79: Secrets library uses Vault code with Business Source License (BUSL-1.1).
The libopenstorage/serets library uses Vault code that is under the business source license (BUSL-1.1). This license prohibits using Hashicorp's code in any "competing products" in production. Hashicorp has kept the definition of what constitutes a competing product vague, meaning that usage of BUSL-1.1 licensed code in production could be high risk. 
More info: https://infisical.com/blog/hashicorp-new-bsl-license
 
This dependency is codified in this repo in the following 2 lines. 
https://github.com/libopenstorage/secrets/blob/efe55db6c349bba256c8b5868af17062a64031b9/vault/utils/utils.go#L16-L17

`github.com/hashicorp/vault/command` is part of the Vault primary application that is BUSL-1.1 licensed. The Vault API (`github.com/hashicorp/vault/api`) is licensed under the open source Mozilla `MPL-2.0` license.

I can make the changes to libopenstorage/secrets to use the API methods for auth (MPL-2.0) rather than the command methods (BUSL-1.1), but I am concerned that this repository does not have many regular contributors. I would like some assurance before I begin this work that someone is interested in reviewing and merging my pull request before I begin.

The alternative is that the project I am working on (Rook - https://github.com/rook/rook) will stop using libopenstorage/secrets and begin using the Vault API directly, which we don't prefer.

I see that @dahuang-purestorage @arivankar-px and @fmilichovsky are the last 3 contributors to successfully have a PR merged, with @piyush-nimbalkar @ggriffiths and @adityadani providing approving reviews. Could any of you provide review and merge support my work, assuming it meets your repo's quality guidelines within the next 2 weeks?

### PR DESCRIPTION
**Pull request by BlaineEXE**
**What this PR does / why we need it**:

Switch usage of Vault libraries that are licensed under Business Source License (BUSL-1.1) in favor of Vault libraries licensed under the the open source Mozilla Public License (MPL-2.0).

In practice, this means avoiding code imported by
`github.com/hashicorp/vault/command` which is part of the top-level Vault application licensed under BUSL-1.1 and instead using code imported by `github.com/hashicorp/vault/api` which is not part of the top-level Vault application and which is licensed under the open source MPL-2.0 license.

**Which issue(s) this PR fixes** (optional)
Resolves #79

**Special notes for your reviewer**:

### CODE REVIEW

**Comment by BlaineEXE** [comment]
Requesting review from @ggriffiths @dahuang-purestorage, and @adityadani, please and thank you.

Currently this is a draft PR while I am also letting Rook's CI test Rook's Vault integration with changes proposed here.

Rook PR updated to use WIP of this branch here: https://github.com/rook/rook/pull/13032

File: vault/utils/utils.go [hunk 1]
```diff
@@ -166,26 +157,31 @@ func GetAuthToken(client *api.Client, config map[string]interface{}) (string, er
...
-		return authKubernetes(client, config)
+func authKubernetes(client *api.Client, config map[string]interface{}) (*api.Secret, error) {
+	role := GetVaultParam(config, AuthKubernetesRole)
+	if role == "" {
+		return nil, ErrKubernetesRole
 	}
-	return "", nil, nil, fmt.Errorf("%s method: %s", method, ErrAuthMethodUnknown)
-}
 
-func authKubernetes(client *api.Client, config map[string]interface{}) (string, http.Header, map[string]interface{}, error) {
```
Comment (BlaineEXE) [inline vault/utils/utils.go:178]: Conveniently, `authAppRole()` below was already using the Vault API, so I was able to update the `authKubernetes()` method here to use the same basic flow.

I kept the same basic code logic from `buildAuthConfig()` for setting values from `VAULT_AUTH_KUBERNETES_ROLE`, `VAULT_AUTH_MOUNT_PATH` and `VAULT_AUTH_KUBERNETES_TOKEN_PATH` in the modified `authKubernetes()` as well to make sure behavior remains the same for default cases.

**Comment by BlaineEXE** [comment]
K8s auth integration passes in Rook's CI here: https://github.com/rook/rook/actions/runs/6474391180/job/17579208709?pr=13032

We've at least been able to vet that these changes don't cause any issues for Rook's usage of Vault secrets integration with K8s. Based on that, I don't think breakage of other consumers is a risk.

**Review by dahuang-purestorage — APPROVED** [review]
LGTM! thanks for the great work @BlaineEXE

**Comment by BlaineEXE** [comment]
Is it normal that the AWS and Google secrets tests fail? They look like issues with setting up the tests rather than failures in the secret library.

**Review by ggriffiths — APPROVED** [review]
Looks good to me. Will ping @adityadani for a review. The test failing is unrelated.

**Review by adityadani — APPROVED** [review]
looks good. Thanks @BlaineEXE for the quick turnaround!

**Comment by BlaineEXE** [comment]
Likewise, thank you!

**Pull request merged**